api/snapshot.go

```diff
@@ -0,0 +1,691 @@
+	defer j.Unlock()
+
+	for _, node := range nodes {
+		url := fmt.Sprintf("http://%s:%d%s/logs", node.IP, config.FlagPort, BaseRoute)
+		select {
+		case _, ok := <-j.cancelChan:
+			if ok {
+				j.Status = "Job has been canceled"
+				j.Errors = append(j.Errors, j.Status)
+				log.Debug(j.Status)
+				os.Remove(zipfile.Name())
+				j.LastSnapshotPath = ""
+				return
+			}
+			errMsg := "cancelChan is closed!"
+			j.Errors = append(j.Errors, errMsg)
+			return
+
+		default:
+			j.Status = fmt.Sprintf("Collecting from a node: %s, url: %s", node.IP, url)
+			log.Debug(j.Status)
+		}
+
+		endpoints := make(map[string]string)
+		body, statusCode, err := DCOSTools.Get(url, time.Duration(time.Second*3))
+		if err != nil {
+			errMsg := fmt.Sprintf("could not get a list of logs, url: %s, status code %d", url, statusCode)
+			j.Errors = append(j.Errors, errMsg)
+			log.Error(err)
+			updateSummaryReport(errMsg, node, err.Error(), summaryErrorsReport)
+			continue
+		}
+		if err = json.Unmarshal(body, &endpoints); err != nil {
+			errMsg := "could not unmarshal a list of logs, url: " + url
+			j.Errors = append(j.Errors, errMsg)
+			log.Error(err)
+			updateSummaryReport(errMsg, node, err.Error(), summaryErrorsReport)
+			continue
+		}
+
+		// add http endpoints
+		err = j.getHTTPAddToZip(node, endpoints, j.LastSnapshotPath, zipWriter, summaryErrorsReport, config, DCOSTools)
+		if err != nil {
+			errMsg := "could not add logs for a node, url: " + url
+			j.Errors = append(j.Errors, errMsg)
+			log.Error(err)
+			updateSummaryReport(errMsg, node, err.Error(), summaryErrorsReport)
+		}
+	}
+	if len(j.Errors) == 0 {
+		j.Status = "Snapshot job sucessfully finished"
+	}
+}
+
+// delete a snapshot
+func (j *SnapshotJob) delete(snapshotName string, config *Config, DCOSTools DCOSHelper) (response snapshotReportResponse, err error) {
+	if !strings.HasPrefix(snapshotName, "snapshot-") || !strings.HasSuffix(snapshotName, ".zip") {
+		return prepareResponseWithErr(http.StatusServiceUnavailable, errors.New("format allowed  snapshot-*.zip"))
+	}
+
+	j.Lock()
+	defer j.Unlock()
+
+	// first try to locate a snapshot on a local disk.
+	snapshotPath := path.Join(config.FlagSnapshotDir, snapshotName)
+	log.Debugf("Trying remove snapshot: %s", snapshotPath)
+	_, err = os.Stat(snapshotPath)
+	if err == nil {
+		if err = os.Remove(snapshotPath); err != nil {
+			return prepareResponseWithErr(http.StatusServiceUnavailable, err)
+		}
+		msg := "Deleted " + snapshotPath
+		log.Infof(msg)
+		return prepareResponseOk(http.StatusOK, msg)
+	}
+
+	node, _, ok, err := j.isSnapshotAvailable(snapshotName, config, DCOSTools)
+	if err != nil {
+		return prepareResponseWithErr(http.StatusServiceUnavailable, err)
+	}
+	if ok {
+		url := fmt.Sprintf("http://%s:%d%s/report/snapshot/delete/%s", node, config.FlagPort, BaseRoute, snapshotName)
+		j.Status = "Attempting to delete a snapshot on a remote host. POST " + url
+		log.Debug(j.Status)
+		timeout := time.Duration(time.Second * 5)
```

Comment:
Did we want a configurable timeout here? This isn't a blocker, but I think eventually the UI might need to configure this on the fly.